Firestore/core/src/firebase/firestore/auth/firebase_credentials_provider_apple.h

```diff
@@ -15,42 +15,92 @@
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_AUTH_FIREBASE_CREDENTIALS_PROVIDER_H_
 
 #import <Foundation/Foundation.h>
 
+#include <memory>
+#include <mutex>
+
+#include "Firestore/core/src/firebase/firestore/auth/credentials_provider.h"
+#include "Firestore/core/src/firebase/firestore/auth/user.h"
+#include "absl/strings/string_view.h"
+
 @class FIRApp;
 
 namespace firebase {
 namespace firestore {
 namespace auth {
 
-class AppImpl {
+/**
+ * `FirebaseCredentialsProvider` uses Firebase Auth via `FIRApp` to get an auth
+ * token.
+ *
+ * NOTE: To simplify the implementation, it requires that you set
+ * `userChangeListener` with a non-`nil` value no more than once and don't call
+ * `getTokenForcingRefresh:` after setting it to `nil`.
+ *
+ * This class must be implemented in a thread-safe manner since it is accessed
+ * from the thread backing our internal worker queue and the callbacks from
+ * FIRAuth will be executed on an arbitrary different thread.
+ *
+ * For non-Apple desktop build, this is right now just a stub.
+ */
+class FirebaseCredentialsProvider : public CredentialsProvider {
  public:
-  AppImpl(FIRApp* app) : app_(app) {
-  }
+  // TODO(zxu123): Provide a ctor to accept the C++ Firebase Games App, which
+  // deals all platforms. Right now, only works for FIRApp*.
+  /**
+   * Initializes a new FirebaseCredentialsProvider.
+   *
+   * @param app The Firebase app from which to get credentials.
+   */
+  FirebaseCredentialsProvider(FIRApp* app);
+
+  void GetToken(bool force_refresh, TokenListener completion) override;
 
-  operator FIRApp*() const {
-    return app_;
-  }
+  void SetUserChangeListener(UserChangeListener listener) override;
+
+  friend class FirebaseCredentialsProviderTest_GetToken_Test;
+  friend class FirebaseCredentialsProviderTest_SetListener_Test;
+  friend class DISABLED_FirebaseCredentialsProviderTest_GetToken_Test;
+  friend class DISABLED_FirebaseCredentialsProviderTest_SetListener_Test;
 
  private:
-  FIRApp* app_;
-};
+  /** Initialize with default app for internal usage such as test. */
+  FirebaseCredentialsProvider();
+
+  static void PlatformDependentTestSetup(const absl::string_view config_path);
+
+  const FIRApp* app_;
+
+  /**
+   * Handle used to stop receiving auth changes once userChangeListener is
+   * removed.
+   */
+  id<NSObject> auth_listener_handle_;
+
+  /** The current user as reported to us via our AuthStateDidChangeListener. */
+  User current_user_;
 
-struct AuthImpl {
-  const FIRApp* app;
+  /**
+   * Counter used to detect if the user changed while a -getTokenForcingRefresh:
+   * request was outstanding.
+   */
+  int user_counter_;
 
-  /** Handle used to stop receiving auth changes once userChangeListener is
-   * removed. */
-  id<NSObject> auth_listener_handle;
+  // Make it static as as it is used in some of the callbacks. Otherwise, we saw
+  // mutex lock failed: Invalid argument.
+  static std::mutex mutex_;
```

Comment:
done. thanks. Wish the compiler can detect this and give warning; will save me a hour or so on a wrong-fix ...

Reply:
sadly, still does not work:
libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument

Making it static also change when its destruction takes place. My hypothesis is that somehow there is use-change event at the end of the program, which calls the callback, which accesses the mutex_, which however is gone (a.k.a. the destruction of auth-provider proceeds the use-change event). Making it static makes sure the access to mutex_ is still valid.

PS: This only happens in xcode test; cmake test runs fine either way.

Reply:
This still seems problematic. Doesn't this just mean we're failing to properly clean up? For example, if our destructor forced deregistration, wouldn't this work?

Reply:
Fixed. There are two types of callbacks in this auth-provider. One is through NSNotification, which is fine and the object has an observer handle, which can be registered in dtor; the other is through auth's global dispatch queue and there is no way to un-register a callback that's already in the queue. The solution I figured out is to let the caller manager `this` and de-allocate it in the caller's dtor through auth's global dispatch queue as well. Since auth's global dispatch queue is serial, which makes `this` valid for the life of all previous callbacks. How does this sound?